server/src/main/java/org/elasticsearch/action/ActionListener.java

```diff
@@ -425,6 +425,28 @@ public boolean equals(Object obj) {
         }
     }
 
+    /**
+     * @return A listener which (if assertions are enabled) wraps around the given delegate and asserts that it is called at least once.
+     */
+    static <Response> ActionListener<Response> assertAtLeastOnce(ActionListener<Response> delegate) {
+        if (Assertions.ENABLED) {
+            return new ActionListenerImplementations.AssertAtLeastOnceActionListener<>(
+                delegate,
+                // We use maybeDieOnAnotherThread here because this will get executed
+                // on the Cleaner thread, which silently swallows Throwable
+                (listener, createdAt) -> ExceptionsHelper.maybeDieOnAnotherThread(
+                    new AssertionError(
+                        "Expected listener "
+                            + delegate
+                            + " to be called at least once, but it was never called. Created:"
+                            + ExceptionsHelper.formatStackTrace(createdAt.getStackTrace())
```

Comment:
The logging in `maybeDieOnAnotherThread` strips off the cause, so I put the created-at stack trace in the exception message. I may be better to set it as the cause of the `AssertionError` but I don't know what that will look like being handled by the uncaught exception handler.